### PR TITLE
Refactor v2

### PR DESCRIPTION
**Pull request by rfjakob**
Reworked acc. to the feeback in https://github.com/pkg/xattr/pull/20 - I think it's quite a bit nicer than the first try.

Tested on Linux, Darwin, FreeBSD.

### CODE REVIEW

**Comment by rfjakob** [comment]
For the `attrListToStrings` problem I went for the copy-paste in Linux and Darwin. Could also be pulled into xattr.go with an extra layer of indirection, but I though it is cleaner like this.

File: .travis.sh [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#!/bin/sh -e
+
+echo "Building for Linux..."
+GOOS=linux   GOARCH=amd64 go build
```
Comment (kuba--) [inline .travis.sh]: Do we need to specify architecture, here?

File: .travis.sh [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+echo "Building for Darwin..."
+GOOS=darwin  GOARCH=amd64 go build
+echo "Building for FreeBSD..."
+GOOS=freebsd GOARCH=amd64 go build
+
+echo "Running tests.."
+uname -a
+go vet
```
Comment (kuba--) [inline .travis.sh]: We can just define `GOOS=darwin` for `go vet`. I think there is no need for `build`.

File: .travis.yml [hunk 3]
```diff
@@ -18,6 +18,5 @@ install:
   - go build
 
 script:
-  - go vet
-  - go test -v -race
+  - ./.travis.sh
```
Comment (kuba--) [inline .travis.yml]: maybe just `sh .travis.sh`

File: xattr_darwin.go [hunk 4]
```diff
@@ -2,72 +2,76 @@
...
-		return nil, &Error{"xattr.Get", path, name, err}
+func getxattr(path string, attr string, dest []byte) (sz int, err error) {
+	value, size := sliceToPtr(dest)
+	/*
+		ssize_t getxattr(const char *path, const char *name,
+			void *value, size_t size, u_int32_t position, int options)
+	*/
+	r0, _, e1 := syscall.Syscall6(syscall.SYS_GETXATTR, uintptr(unsafe.Pointer(syscall.StringBytePtr(path))),
+		uintptr(unsafe.Pointer(syscall.StringBytePtr(attr))), uintptr(unsafe.Pointer(value)), uintptr(size), 0, 0)
+	if e1 != 0 {
```
Comment (kuba--) [inline xattr_darwin.go]: lets keep verbose `e1 != syscall.Errno(0) `

File: xattr_darwin.go [hunk 5]
```diff
@@ -2,72 +2,76 @@
...
+	return int(r0), nil
 }
 
-// Supported checks if filesystem supports extended attributes
-func Supported(path string) bool {
-	if _, err := listxattr(path, nil, 0, 0); err != nil {
-		return err != syscall.ENOTSUP
+// attrListToStrings converts a sequence of attribute name entries to a
+// []string.
+// On Darwin and Linux, each entry is a NUL-terminated string.
```
Comment (kuba--) [inline xattr_darwin.go]: ...an array of NULL terminated UTF-8 strings.

File: xattr_darwin.go (hunk 4 above)
Comment (rfjakob) [reply xattr_darwin.go]: As you prefer, but note that stdlib does it like this
```go
	_, _, e1 := Syscall(SYS_REMOVEXATTR, uintptr(unsafe.Pointer(_p0)), uintptr(unsafe.Pointer(_p1)), 0)
	if e1 != 0 {
		err = errnoErr(e1)
	}
```
https://github.com/golang/go/blob/dc3a92edaabf012cabd90304bac1ed8fe3bb7444/src/syscall/zsyscall_linux_amd64.go#L771

File: .travis.sh (hunk 1 above)
Comment (rfjakob) [reply .travis.sh]: Actually, no. Will drop that.

File: .travis.sh (hunk 2 above)
Comment (rfjakob) [reply .travis.sh]: Hmm I don't think we should assume that `go vet` runs the same checks that `go build` does. Or is this documented?

File: .travis.yml (hunk 3 above)
Comment (rfjakob) [reply .travis.yml]: Ok

File: xattr_darwin.go (hunk 5 above)
Comment (rfjakob) [reply xattr_darwin.go]: Ok

File: .travis.sh (hunk 2 above)
Comment (rfjakob) [reply .travis.sh]: I think this is pretty clear - we cannot rely on it:
> Go vet never builds, so any try build-time flags are pointless. The issue is how the go command identifies packages to pass to go tool vet.
https://github.com/golang/go/issues/21711#issuecomment-326197693

Comment (rfjakob) [reply .travis.sh]: We should do both, build and vet

Comment (kuba--) [reply .travis.sh]: sounds good

File: xattr_darwin.go (hunk 4 above)
Comment (kuba--) [reply xattr_darwin.go]: Hard coded `0` looks like a magic number. Moreover the file was autogenerated by:
`mksyscall.pl -tags linux,amd64 syscall_linux.go syscall_linux_amd64.go`

**Comment by kuba--** [comment]
the same changes are already in  https://github.com/pkg/xattr/pull/24

**Pull request closed without merging**